### PR TITLE
refactor: AGNT redemption unleash

### PR DESCRIPTION
**Pull request by kupermind**
- AGNT redemption unleash;
- Audit fixes.

### CODE REVIEW

File: contracts/MemeFactory.sol [hunk 1]
```diff
@@ -18,8 +18,8 @@ interface IWETH {
 
 // UniswapV2 interface
 interface IUniswap {
-    /// @dev Creates an LP pair.
-    function createPair(address tokenA, address tokenB) external returns (address pair);
```
Comment (kupermind) [inline contracts/MemeFactory.sol:22]: addLiquidity() creates a pair already if it does not exist.

File: contracts/MemeFactory.sol [hunk 2]
```diff
@@ -60,7 +60,7 @@ abstract contract MemeFactory {
     event Summoned(address indexed summoner, address indexed memeToken, uint256 nativeTokenContributed);
     event Hearted(address indexed hearter, address indexed memeToken, uint256 amount);
     event Unleashed(address indexed unleasher, address indexed memeToken, address indexed lpPairAddress,
-        uint256 liquidity, uint256  burnPercentageOfStable);
```
Comment (kupermind) [inline contracts/MemeFactory.sol:63]: Better naming

File: contracts/MemeFactory.sol [hunk 3]
```diff
@@ -92,22 +92,32 @@ abstract contract MemeFactory {
     address public constant OLAS_BURNER = 0x51eb65012ca5cEB07320c497F4151aC207FEa4E0;
     // Meme token decimals
     uint8 public constant DECIMALS = 18;
+    // AGNT redemption amount as per:
+    // https://basescan.org/address/0x42156841253f428cb644ea1230d4fddfb70f8891#readContract#F17
+    // Previous token address: 0x7484a9fB40b16c4DFE9195Da399e808aa45E9BB9
+    // Full collected amount: 141569842100000000000
+    // Redemption amount: collected amount - 10% for burn = 127412857890000000000
+    uint256 public constant REDEMPTION_AMOUNT = 127412857890000000000;
```
Comment (kupermind) [inline contracts/MemeFactory.sol:100]: Please check math

File: contracts/MemeFactory.sol [hunk 4]
```diff
@@ -192,6 +200,9 @@ abstract contract MemeFactory {
             address(this),
             block.timestamp
         );
+
+        // Get the pair address
+        pair = IUniswap(factory).getPair(nativeToken, memeToken);
```
Comment (kupermind) [inline contracts/MemeFactory.sol:205]: We add liquidity and then just get the pair address for return.

File: contracts/MemeFactory.sol [hunk 5]
```diff
@@ -271,7 +330,7 @@ abstract contract MemeFactory {
         uint256 totalNativeTokenCommitted = memeSummon.nativeTokenContributed;
 
         // Check that the meme has been summoned
-        require(totalNativeTokenCommitted > 0, "Meme not yet summoned");
```
Comment (kupermind) [inline contracts/MemeFactory.sol:274]: This change will prevent hearing the redemption token.

File: contracts/MemeFactory.sol [hunk 6]
```diff
@@ -298,26 +357,52 @@ abstract contract MemeFactory {
         // Get the total native token amount committed to this meme
         uint256 totalNativeTokenCommitted = memeSummon.nativeTokenContributed;
 
+        // Check if the meme has been summoned
+        require(memeSummon.unleashTime == 0, "Meme already unleashed");
```
Comment (kupermind) [inline contracts/MemeFactory.sol:361]: This is needed not to have a double unleash going too far.

File: contracts/MemeFactory.sol [hunk 7]
```diff
@@ -298,26 +357,52 @@ abstract contract MemeFactory {
...
+            }
+        }
+
+        // Schedule native token amount for ascendance
+        scheduledForAscendance += nativeAmountForOLASBurn;
 
         // Calculate LP token allocation according to LP percentage and distribution to supporters
         Meme memeTokenInstance = Meme(memeToken);
         uint256 totalSupply = memeTokenInstance.totalSupply();
-        uint256 lpTokenAmount = (totalSupply * LP_PERCENTAGE) / 100;
```
Comment (kupermind) [inline contracts/MemeFactory.sol:316]: Better naming

**Review by DavidMinarsch — APPROVED** [review]

File: contracts/MemeFactory.sol (hunk 5 above)
Comment (DavidMinarsch) [reply contracts/MemeFactory.sol:274]: Check myself

File: contracts/MemeFactory.sol (hunk 3 above)
Comment (DavidMinarsch) [reply contracts/MemeFactory.sol]: Check myself

**Pull request merged**